### PR TITLE
feat: Add SpeculativeDecodingTrainer with TRAIN_ONLY and DATA_ONLY mode support 

### PR DESCRIPTION
**Pull request by hrathina**
## Summary

Adds `SpeculativeDecodingTrainer` to the Kubeflow SDK that wraps the [speculators](https://github.com/vllm-project/speculators) library to train Eagle3 draft models for speculative decoding inference acceleration on Kubernetes.

- New trainer: `SpeculativeDecodingTrainer` with `SpeculatorConfig`
- Supports Eagle3 speculator architecture
- 2 modes: `train_only`, `data_only`
- Managed vLLM sidecar for hidden state extraction (DATA_ONLY)
- Multi-GPU training via `torchrun` with NCCL (TRAIN_ONLY)
- Progression tracking with HTTP metrics server for both modes
- Auto layer detection from verifier model config
- PVC URI support (`pvc://<name>/<path>`) for all paths
- Compatible with speculators 0.6.0 and vLLM 0.24

## How It Works

Speculative decoding trains a small "draft" model that predicts tokens using hidden states from intermediate layers of a large "verifier" model. The two stages can run independently as separate Kubernetes jobs:

1. **Data extraction (DATA_ONLY)** — preprocess dataset, optionally regenerate responses, extract hidden states from verifier model via vLLM sidecar
2. **Training (TRAIN_ONLY)** — build vocab mappings, create EAGLE3 draft model, train on pre-extracted hidden states

### DATA_ONLY — extract hidden states via managed vLLM sidecar
```python
from kubeflow.trainer import TrainerClient
from kubeflow.trainer.rhai.speculator import (
    SpeculativeDecodingTrainer,
    SpeculatorMode,
    SpeculatorConfig,
)

client = TrainerClient()
client.train(
    name="eagle3-data-job",
    trainer=SpeculativeDecodingTrainer(
        mode=SpeculatorMode.DATA_ONLY,
        verifier_model="Qwen/Qwen3-8B",
        dataset_name="sharegpt",
        output_dir="pvc://shared/speculator/my-run",
        max_samples=5000,
        total_seq_len=512,
        vllm_gpu_count=1,
        regenerate_responses=True,
    ),
    runtime_ref="speculator-data-extract-cuda",
)
```

### TRAIN_ONLY — train from pre-extracted hidden states on PVC
```python
client.train(
    name="eagle3-train-job",
    trainer=SpeculativeDecodingTrainer(
        mode=SpeculatorMode.TRAIN_ONLY,
        verifier_model="Qwen/Qwen3-8B",
        data_path="pvc://shared/speculator/my-run",
        hidden_states_path="pvc://shared/speculator/my-run/hidden_states",
        output_dir="pvc://shared/speculator/my-run/checkpoints",
        training_gpu_count=2,
        epochs=3,
        lr=1e-4,
        config=SpeculatorConfig(
            resume_from_checkpoint=True,
        ),
    ),
    runtime_ref="speculator-train-only-cuda",
)
```

### DATA_ONLY with PVC model (pre-downloaded)
```python
client.train(
    name="eagle3-data-pvc",
    trainer=SpeculativeDecodingTrainer(
        mode=SpeculatorMode.DATA_ONLY,
        verifier_model="pvc://shared/models/Qwen3-8B",
        dataset_name="sharegpt",
        output_dir="pvc://shared/speculator/my-run",
        config=SpeculatorConfig(
            target_layer_ids=[2, 18, 33, 36],  # required for PVC models
        ),
    ),
    runtime_ref="speculator-data-extract-cuda",
)
```

### TRAIN_ONLY with PVC model (pre-downloaded)
```python
client.train(
    name="eagle3-train-pvc",
    trainer=SpeculativeDecodingTrainer(
        mode=SpeculatorMode.TRAIN_ONLY,
        verifier_model="pvc://shared/models/Qwen3-8B",
        data_path="pvc://shared/speculator/my-run",
        hidden_states_path="pvc://shared/speculator/my-run/hidden_states",
        output_dir="pvc://shared/speculator/my-run/checkpoints",
        training_gpu_count=2,
        epochs=3,
        lr=1e-4,
        config=SpeculatorConfig(
            target_layer_ids=[2, 18, 33, 36],  # required for PVC models
        ),
    ),
    runtime_ref="speculator-train-only-cuda",
)
```

## vLLM Sidecar (DATA_ONLY)

The vLLM sidecar runs as an init container within the same pod. It serves the verifier model for hidden state extraction.

| Parameter | Default | Description |
|-----------|---------|-------------|
| `vllm_gpu_count` | `1` | Number of GPUs for vLLM (uses `--tensor-parallel-size` when > 1) |
| `vllm_gpu_memory_utilization` | `0.9` | Fraction of GPU memory for vLLM |

### SpeculativeDecodingTrainer

| Parameter | Default | Description |
|-----------|---------|-------------|
| `verifier_model` | required | HF model ID or `pvc://<name>/<path>` |
| `mode` | required | `TRAIN_ONLY` or `DATA_ONLY` |
| `speculator_type` | `EAGLE3` | Draft model architecture |
| `dataset_name` | — | Dataset for extraction: `sharegpt`, `ultrachat`, `gsm8k`, or HF ID (DATA_ONLY) |
| `data_path` | — | PVC URI to preprocessed Arrow dataset (TRAIN_ONLY) |
| `hidden_states_path` | — | PVC URI to extracted hidden states (TRAIN_ONLY) |
| `output_dir` | — | PVC URI for outputs |
| `max_samples` | — | Max dataset samples |
| `epochs` | `3` | Training epochs |
| `lr` | `1e-4` | Learning rate |
| `total_seq_len` | `2048` | Maximum sequence length for preprocessing, vLLM context window, and training |
| `draft_vocab_size` | auto | Vocab size for draft model (auto: `min(8192, verifier_vocab_size)`) |
| `training_gpu_count` | `1` | GPUs for training |
| `vllm_gpu_count` | `1` | GPUs for vLLM sidecar |
| `vllm_gpu_memory_utilization` | `0.9` | Fraction of GPU memory for vLLM |
| `config` | — | Advanced training configuration (`SpeculatorConfig`) |
| `packages_to_install` | — | Python packages to install in pod |
| `pip_index_urls` | defaults | PyPI index URLs for package installation |
| `env` | — | Environment variables for training pods |
| `regenerate_responses` | `False` | Regenerate responses from verifier before extraction |
| `vllm_endpoint` | — | URL of user-managed vLLM endpoint (OFFLINE mode) |
| `enable_progression_tracking` | `True` | Enable progression tracking |
| `metrics_port` | `28080` | HTTP server port for metrics endpoint |
| `metrics_poll_interval_seconds` | `30` | How often controller polls metrics (seconds) |

### SpeculatorConfig

| Parameter | Default | Description |
|-----------|---------|-------------|
| `num_layers` | `1` | Draft model layers |
| `ttt_steps` | `3` | Test-time training steps |
| `norm_before_residual` | `True` | Apply normalization before residual connection |
| `norm_before_fc` | `False` | Apply normalization before fully-connected layer |
| `embed_requires_grad` | `False` | Whether embedding layer requires gradients |
| `hidden_states_dtype` | `bfloat16` | Dtype for hidden states (`bfloat16`, `float16`, `float32`) |
| `scheduler_type` | `linear` | LR scheduler (`linear`, `cosine`, `none`) |
| `scheduler_warmup_steps` | auto | Warmup steps (default: 1% of total training steps) |
| `scheduler_total_steps` | auto | Total scheduler steps (default: `num_epochs * steps_per_epoch`) |
| `scheduler_num_cosine_cycles` | `0.5` | Cosine cycles (only for `cosine` scheduler) |
| `checkpoint_freq` | `1.0` | Checkpoint frequency in epochs |
| `save_best` | `False` | Save best model by validation loss |
| `log_freq` | `1` | Logging frequency in steps |
| `resume_from_checkpoint` | `False` | Resume from existing checkpoint |
| `datagen_concurrency` | `4` | Concurrent vLLM requests for extraction |
| `target_layer_ids` | auto | Layer IDs for hidden state extraction (auto-detected for HF models) |

## Output

Checkpoints are saved by the speculators `Trainer`:
```
checkpoints/
  0/                      # epoch 0 (full state: model + optimizer + scheduler)
  1/                      # epoch 1
  checkpoint_best/        # best by validation loss (if save_best=True)
  interrupted/            # model-only checkpoint on interruption (auto-deleted on resume)
```

## Test plan

Tested with speculators 0.6.0 + vLLM 0.24 on Qwen/Qwen3-8B:

**TRAIN_ONLY:**
- [x] Fresh training with 2 GPUs — distributed training with NCCL, vocab mapping sync across ranks
- [x] Resume from checkpoint — resumes from latest epoch, progression tracking accounts for completed epochs
- [x] Resume with interrupted checkpoint — `interrupted` directory auto-deleted by rank 0 before resume
- [x] Multi-GPU barrier synchronization — rank 0 creates vocab mappings, all ranks wait via barrier
- [x] Progression tracking — correct percentage for fresh and resumed runs
- [x] HuggingFace model with auto layer detection
- [x] PVC model with explicit `target_layer_ids`

**DATA_ONLY:**
- [x] End-to-end extraction with vLLM sidecar (single GPU)
- [x] Response regeneration (`regenerate_responses=True`)
- [x] Already-completed extraction skip — skips when hidden states exist
- [x] Progression tracking — file-counting based progress

**Unit tests:**
- [x] 94 tests covering script rendering, parameter validation, mode-specific logic, and progression instrumentation

**Follow-up:** E2E automation tests will be created after this PR is merged as an immediate follow-up.

### CODE REVIEW

**Comment by efazal** [comment]
Thanks @hrathina. Good progress since the last round — all previous findings addressed. Left a few more comments, mainly around the layer ID formula alignment with speculators upstream.

File: kubeflow/trainer/rhai/speculator.py [hunk 1]
```diff
@@ -0,0 +1,1550 @@
...
+                            f"verifier_model {self.verifier_model!r} is not a valid "
+                            f"HuggingFace model ID. For DATA_ONLY mode, verifier_model "
+                            f"must be either a HuggingFace model ID "
+                            f"(e.g. 'meta-llama/Llama-3.1-8B-Instruct') or a PVC URI "
+                            f"(pvc://<name>/<path>)."
+                        ) from e
+                    if hasattr(model_config, "text_config"):
+                        model_config = model_config.text_config
+                    n = model_config.num_hidden_layers
+                    cfg.target_layer_ids = [2, n // 2, n - 3, n]
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:491]: SDK auto-detects `[2, n//2, n-3, n]` (4 layers) but speculators' `resolve_target_layer_ids` defaults to `[2, n//2, n-3]` (3 layers). A user running DATA_ONLY then TRAIN_ONLY without explicit `target_layer_ids` gets a shape mismatch: 4-layer extracted data vs 3-layer model.

File: kubeflow/trainer/rhai/speculator.py [hunk 2]
```diff
@@ -0,0 +1,1550 @@
...
+        "verifier_name_or_path": verifier_model,
+        "d2t": d2t,
+        "t2d": t2d,
+    }
+    if target_layer_ids is not None:
+        from_training_kwargs["target_layer_ids"] = target_layer_ids
+    model = Eagle3DraftModel.from_training_args(verifier_config, **from_training_kwargs)
+
+    max_len = total_seq_len
+    hs_dtype = getattr(torch, hidden_states_dtype)
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:823]: nit: missing `num_target_layers` param (defaults to 3). If `target_layer_ids` has a different count, the empty-batch fallback creates wrong-shaped tensors. Pass `num_target_layers=len(target_layer_ids) if target_layer_ids else 3`.

File: kubeflow/trainer/rhai/speculator.py [hunk 3]
```diff
@@ -0,0 +1,1550 @@
...
+    if os.path.exists(incomplete_marker):
+        print(
+            f"[Kubeflow] Warning: Previous data extraction for rank {rank} did not complete. "
+            "Restarting extraction from the beginning.",
+            flush=True,
+        )
+    elif any(f.endswith(".safetensors") for f in os.listdir(hidden_states_dir)):
+        print("[Kubeflow] Data extraction already completed. Skipping.", flush=True)
+        if "_mark_data_complete" in globals():
+            _mark_data_complete()  # noqa: F821
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:548]: nit: early-exit path calls `_mark_data_complete()` but never writes the termination message to `/dev/termination-log`. The dashboard will show the job stuck below 100%.

File: kubeflow/trainer/rhai/speculator.py [hunk 4]
```diff
@@ -0,0 +1,1550 @@
...
+            "Re-run data extraction (DATA_ONLY mode) before training."
+        )
+
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    is_distributed = int(os.environ.get("WORLD_SIZE", 1)) > 1
+    if is_distributed and not torch.distributed.is_initialized():
+        torch.distributed.init_process_group(backend="nccl")
+        torch.cuda.set_device(local_rank)
+
+    verifier_config = AutoConfig.from_pretrained(verifier_model)
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: Missing `token=os.environ.get("HF_TOKEN")` for gated models. The validation path at line 428 passes the token but this runtime call does not. Gated models will 401 at training time.

File: kubeflow/trainer/rhai/speculator.py [hunk 5]
```diff
@@ -0,0 +1,1550 @@
...
+
+    num_layers: int = 1
+    ttt_steps: int = 3
+    norm_before_residual: bool = True
+    norm_before_fc: bool = False
+    embed_requires_grad: bool = False
+    hidden_states_dtype: str = "bfloat16"
+    scheduler_type: str = "linear"
+    scheduler_warmup_steps: int | None = None
+    scheduler_total_steps: int | None = None
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:107]: nit: no init-time validation for `scheduler_type`. Invalid values only fail at runtime inside the training pod. Consider validating against `('linear', 'cosine', 'none')` here.

File: kubeflow/trainer/rhai/speculator_test.py [hunk 6]
```diff
@@ -0,0 +1,1720 @@
...
+        mode=SpeculatorMode.TRAIN_ONLY,
+        training_resources={"nvidia.com/gpu": 1},
+        hidden_states_path="pvc://test-pvc/hidden_states",
+        data_path="pvc://test-pvc/arrow_dataset",
+        output_dir="pvc://test-pvc/output",
+        epochs=5,
+        lr=1e-5,
+    )
+
+    script = _render_speculator_training_script(trainer)
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator_test.py:340]: nit: adding `compile(script, '<test>', 'exec')` to the script content tests would catch syntax errors in rendered code before pod submission.

File: kubeflow/trainer/rhai/speculator.py (hunk 1 above)
Comment (hrathina) [reply kubeflow/trainer/rhai/speculator.py:491]: I faced this issue during testing — Eagle3 training requires 4 layers (3 as input to the draft model + 1 to recompute the target distribution). When I used 3 layers, it caused a shape mismatch during data extraction and training. I hit the same issue in ONLINE mode as well. vLLM's own extraction examples also use 4 layers (e.g. [3, 18, 33, 36]). The speculators resolve_target_layer_ids defaulting to 3 is for inference, not training. That's why I set 4 layers [2, n//2, n-3, n] as the default.

**Comment by hrathina** [comment]
> How do we specify other vLLM params? Are we only exposing these two at the moment?
> 
> ```
> vllm_gpu_count | 1 | Number of GPUs for vLLM (uses --tensor-parallel-size when > 1)
> vllm_gpu_memory_utilization | 0.9 | Fraction of GPU memory for vLLM
> ```

Yes, currently we only expose vllm_resources and vllm_gpu_memory_utilization (we've updated vllm_gpu_count to vllm_resources). Right now we can't update the sidecar YAML directly. From the next Trainer 2.3 version we'll support patching the sidecar with RuntimePatch, which will allow exposing additional vLLM params.

```
vllm_resources={"nvidia.com/gpu": 1, "memory": "64Gi", "cpu": "1"},
vllm_gpu_memory_utilization=0.9,
```

**Comment by efazal** [comment]
Thanks @hrathina. Great work!
/lgtm
